### ISSUE TITLE
UserResolver stoped working on version 13.x

### ISSUE THREAD

**Issue by Messhias** [owen-it/laravel-auditing#694]
| Q                 | A
| ----------------- | ---
| Bug?              | yes
| New Feature?      | no
| Framework         | Laravel
| Framework version | 8.8.x
| Package version   | 13
| PHP version       | 8.1.2

#### Actual Behaviour

With the new update I got this:
```
[2022-03-03 15:28:22] local.ERROR: OwenIt\Auditing\Exceptions\AuditingException: Invalid UserResolver implementation in /Users/collegia/Documents/Collegia/systems/api/vendor/owen-it/laravel-auditing/src/Auditable.php:363
Stack trace:
#0 /Users/collegia/Documents/Collegia/systems/api/vendor/owen-it/laravel-auditing/src/Auditable.php(326): App\Models\Base->resolveUser()
#1 /Users/collegia/Documents/Collegia/systems/api/vendor/owen-it/laravel-auditing/src/Drivers/Database.php(19): App\Models\Base->toAudit()
#2 /Users/collegia/Documents/Collegia/systems/api/vendor/owen-it/laravel-auditing/src/Auditor.php(81): OwenIt\Auditing\Drivers\Database->audit(Object(App\Models\DealOwner))
#3 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php(261): OwenIt\Auditing\Auditor->execute(Object(App\Models\DealOwner))
#4 /Users/collegia/Documents/Collegia/systems/api/vendor/owen-it/laravel-auditing/src/AuditableObserver.php(38): Illuminate\Support\Facades\Facade::__callStatic('execute', Array)
#5 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(424): OwenIt\Auditing\AuditableObserver->created(Object(App\Models\DealOwner))
#6 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(249): Illuminate\Events\Dispatcher->Illuminate\Events\{closure}('eloquent.create...', Array)
#7 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Concerns/HasEvents.php(189): Illuminate\Events\Dispatcher->dispatch('eloquent.create...', Array)
#8 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1174): Illuminate\Database\Eloquent\Model->fireModelEvent('created', false)
#9 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(994): Illuminate\Database\Eloquent\Model->performInsert(Object(Illuminate\Database\Eloquent\Builder))
#10 /Users/collegia/Documents/Collegia/systems/api/app/Repositories/RepositoryEloquent.php(277): Illuminate\Database\Eloquent\Model->save()
#11 /Users/collegia/Documents/Collegia/systems/api/app/Repositories/RepositoryEloquent.php(254): App\Repositories\RepositoryEloquent->saveObj(Array, true)
#12 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Database/Concerns/ManagesTransactions.php(29): App\Repositories\RepositoryEloquent->App\Repositories\{closure}(Object(App\Support\PostgresConnection))
#13 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Database/DatabaseManager.php(442): Illuminate\Database\Connection->transaction(Object(Closure))
#14 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php(261): Illuminate\Database\DatabaseManager->__call('transaction', Array)
#15 /Users/collegia/Documents/Collegia/systems/api/app/Repositories/RepositoryEloquent.php(255): Illuminate\Support\Facades\Facade::__callStatic('transaction', Array)
#16 /Users/collegia/Documents/Collegia/systems/api/app/Http/Controllers/ResourceApiController.php(92): App\Repositories\RepositoryEloquent->create(Array)
#17 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Routing/Controller.php(54): App\Http\Controllers\ResourceApiController->create(Object(Illuminate\Http\Request))
#18 /Users/collegia/Documents/Collegia/systems/api/app/Http/Controllers/ResourceApiController.php(48): Illuminate\Routing\Controller->callAction('create', Array)
#19 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(45): App\Http\Controllers\ResourceApiController->callAction('create', Array)
#20 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Routing/Route.php(262): Illuminate\Routing\ControllerDispatcher->dispatch(Object(Illuminate\Routing\Route), Object(App\Http\Controllers\API\Admin\DealOwnersController), 'create')
#21 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Routing/Route.php(205): Illuminate\Routing\Route->runController()
#22 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Routing/Router.php(721): Illuminate\Routing\Route->run()
#23 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\Routing\Router->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#24 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php(50): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#25 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\Routing\Middleware\SubstituteBindings->handle(Object(Illuminate\Http\Request), Object(Closure))
#26 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Routing/Middleware/ThrottleRequests.php(127): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#27 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Routing/Middleware/ThrottleRequests.php(63): Illuminate\Routing\Middleware\ThrottleRequests->handleRequest(Object(Illuminate\Http\Request), Object(Closure), Array)
#28 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\Routing\Middleware\ThrottleRequests->handle(Object(Illuminate\Http\Request), Object(Closure), '60', '1')
#29 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#30 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Routing/Router.php(723): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#31 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Routing/Router.php(698): Illuminate\Routing\Router->runRouteWithinStack(Object(Illuminate\Routing\Route), Object(Illuminate\Http\Request))
#32 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Routing/Router.php(662): Illuminate\Routing\Router->runRoute(Object(Illuminate\Http\Request), Object(Illuminate\Routing\Route))
#33 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Routing/Router.php(651): Illuminate\Routing\Router->dispatchToRoute(Object(Illuminate\Http\Request))
#34 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(167): Illuminate\Routing\Router->dispatch(Object(Illuminate\Http\Request))
#35 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\Foundation\Http\Kernel->Illuminate\Foundation\Http\{closure}(Object(Illuminate\Http\Request))
#36 /Users/collegia/Documents/Collegia/systems/api/app/Http/Middleware/CorsMiddleware.php(30): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#37 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): App\Http\Middleware\CorsMiddleware->handle(Object(Illuminate\Http\Request), Object(Closure))
#38 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#39 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ConvertEmptyStringsToNull.php(31): Illuminate\Foundation\Http\Middleware\TransformsRequest->handle(Object(Illuminate\Http\Request), Object(Closure))
#40 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull->handle(Object(Illuminate\Http\Request), Object(Closure))
#41 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#42 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TrimStrings.php(40): Illuminate\Foundation\Http\Middleware\TransformsRequest->handle(Object(Illuminate\Http\Request), Object(Closure))
#43 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\Foundation\Http\Middleware\TrimStrings->handle(Object(Illuminate\Http\Request), Object(Closure))
#44 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php(27): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#45 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\Foundation\Http\Middleware\ValidatePostSize->handle(Object(Illuminate\Http\Request), Object(Closure))
#46 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php(86): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#47 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\Foundation\Http\Middleware\PreventRequestsDuringMaintenance->handle(Object(Illuminate\Http\Request), Object(Closure))
#48 /Users/collegia/Documents/Collegia/systems/api/vendor/fruitcake/laravel-cors/src/HandleCors.php(38): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#49 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Fruitcake\Cors\HandleCors->handle(Object(Illuminate\Http\Request), Object(Closure))
#50 /Users/collegia/Documents/Collegia/systems/api/vendor/fideloper/proxy/src/TrustProxies.php(57): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#51 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Fideloper\Proxy\TrustProxies->handle(Object(Illuminate\Http\Request), Object(Closure))
#52 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#53 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(142): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#54 /Users/collegia/Documents/Collegia/systems/api/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(111): Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter(Object(Illuminate\Http\Request))
#55 /Users/collegia/Documents/Collegia/systems/api/public/index.php(55): Illuminate\Foundation\Http\Kernel->handle(Object(Illuminate\Http\Request))
#56 /Users/collegia/Documents/Collegia/systems/api/server.php(21): require_once('/Users/collegia...')
#57 {main}  
```

#### Expected Behaviour

Work normally.

#### Steps to Reproduce

Use this `UserResolver` class:

```

/**
 * Class UserResolver
 * @package App\Resolvers
 */
class UserResolver implements \OwenIt\Auditing\Contracts\UserResolver
{
	/**
	 * {@inheritdoc}
	 */
	public static function resolve()
	{
		$guards = config('audit.user.guards');
		foreach ($guards as $guard) {
			if (Auth::guard($guard)->check()) {
				return Auth::guard($guard)->user();
			}
		}
		
		return null;
	}
}
```

**Comment by MortenDHansen** [comment]
Please show me the config. I can't reproduce the issue so i need to know more 🤷‍♂️

**Comment by Cigoler** [comment]
Sorry for being dense, but what do you mean by updating the config? This is my config:

```
<?php

return [

    'enabled' => env('AUDITING_ENABLED', true),

    /*
    |--------------------------------------------------------------------------
    | Audit Implementation
    |--------------------------------------------------------------------------
    |
    | Define which Audit model implementation should be used.
    |
    */

    'implementation' => OwenIt\Auditing\Models\Audit::class,

    /*
    |--------------------------------------------------------------------------
    | User Morph prefix & Guards
    |--------------------------------------------------------------------------
    |
    | Define the morph prefix and authentication guards for the User resolver.
    |
    */

    'user' => [
        'morph_prefix' => 'user',
        'guards'       => [
            'web',
            'api',
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Audit Resolvers
    |--------------------------------------------------------------------------
    |
    | Define the User, IP Address, User Agent and URL resolver implementations.
    |
    */
    'resolver' => [
        'user'       => OwenIt\Auditing\Resolvers\UserResolver::class,
        'ip_address' => OwenIt\Auditing\Resolvers\IpAddressResolver::class,
        'user_agent' => OwenIt\Auditing\Resolvers\UserAgentResolver::class,
        'url'        => OwenIt\Auditing\Resolvers\UrlResolver::class,
    ],

    /*
    |--------------------------------------------------------------------------
    | Audit Events
    |--------------------------------------------------------------------------
    |
    | The Eloquent events that trigger an Audit.
    |
    */

    'events' => [
        'created',
        'updated',
        'deleted',
        'restored',
    ],

    /*
    |--------------------------------------------------------------------------
    | Strict Mode
    |--------------------------------------------------------------------------
    |
    | Enable the strict mode when auditing?
    |
    */

    'strict' => false,

    /*
    |--------------------------------------------------------------------------
    | Audit Timestamps
    |--------------------------------------------------------------------------
    |
    | Should the created_at, updated_at and deleted_at timestamps be audited?
    |
    */

    'timestamps' => false,

    /*
    |--------------------------------------------------------------------------
    | Audit Threshold
    |--------------------------------------------------------------------------
    |
    | Specify a threshold for the amount of Audit records a model can have.
    | Zero means no limit.
    |
    */

    'threshold' => 0,

    /*
    |--------------------------------------------------------------------------
    | Audit Driver
    |--------------------------------------------------------------------------
    |
    | The default audit driver used to keep track of changes.
    |
    */

    'driver' => 'database',

    /*
    |--------------------------------------------------------------------------
    | Audit Driver Configurations
    |--------------------------------------------------------------------------
    |
    | Available audit drivers and respective configurations.
    |
    */

    'drivers' => [
        'database' => [
            'table'      => 'audits',
            'connection' => null,
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Audit Console
    |--------------------------------------------------------------------------
    |
    | Whether console events should be audited (eg. php artisan db:seed).
    |
    */

    'console' => false,
];

```

What am I missing, do I *need* to add that code block to do with empty audits, it sounds like it's optional from the upgrade guide?

**Comment by MortenDHansen** [comment]
> I'll look into it. Please confirm you have updated the config: https://www.laravel-auditing.com/docs/13.0/upgrading

**Comment by MortenDHansen** [comment]
There are changes to the config between version < 13 and 13. 

1. Move the `OwenIt\Auditing\Resolvers\UserResolver::class` to the user config key `resolver => OwenIt\Auditing\Resolvers\UserResolver::class`
2. Rename the resolver config key to resolvers.

Please check the upgrade instructions when upgrading packages major versions as there are probably breaking changes. (Not just this package, but ANY package 🙂 )

I am working on a fix to give a better error message or provide a fallback 🙂

**Comment by Cigoler** [comment]
Okay Morten, I am sorry, I see that's my fault entirely, but I think the documentation isn't super clear unless you *know* that you're doing something wrong. I've not messed with the config from day one, so when it starts talking about custom resolver implementations I glazed over to be honest.

I would perhaps rephrase it to say there are breaking changes to the config which requires everyone to update their resolvers (etc) to the new implementation (or something similar) and make that stand out. Apologies dude.

**Comment by MortenDHansen** [comment]
@Cigoler no problem at all 🙂  - I'm just happy someone takes the time to report so we can improve 🙂 . I'm probably going to make a more explicit error message and update the instructions 👍

**Comment by poldelgado** [comment]
> > I'll look into it. Please confirm you have updated the config: https://www.laravel-auditing.com/docs/13.0/upgrading
> 
> Still persist the issue.

Thank you very much, following the steps of the update guide was the solution for me.
